### PR TITLE
Add custom action support for models already exposed in api that need it

### PR DESCRIPTION
**Pull request by d-m-u**
Adds model custom action for models that support custom buttons. 

In wip because it requires UI components to be added to all these as well

**Related to** 
Host: 
https://github.com/eclarizio/manageiq/commit/8ca58580d4ef1175d68e5871cd7d80678cec1ca0
https://github.com/eclarizio/manageiq-ui-classic/commit/f7d05b4ad8a47951e9781e84234a728953159688
https://github.com/eclarizio/manageiq-api/commit/e6376161bd8db90ba46d6bb6d414e93b7ca9ba2d

Everything else:
https://github.com/ManageIQ/manageiq/pull/16369

### CODE REVIEW

**Comment by abellotti** [comment]
@d-m-u needs specs for those, maybe enhance custom_actions_spec.rb to handle all these updated collections in a common way for custom actions ?

**Comment by abellotti** [comment]
@d-m-u see the tests I wrote here for availability_zones custom actions:
https://github.com/abellotti/manageiq-api/commit/ebf4e588d22ae6c95b6480ba771ca3f8b92fff80

You can use that file and extend it similarly for all other collections we're exposing the custom actions for in this PR.

Thanks,

/cc'ing @jntullo & @gtanzillo

**Comment by d-m-u** [comment]
@miq-bot assign @gmcculloug 
@miq-bot remove_label wip

**Comment by d-m-u** [comment]
@miq-bot unassign @gmcculloug

**Comment by gmcculloug** [comment]
@gtanzillo @abellotti Who should this be assigned to?  (Hint: not me 🙏 )

File: spec/requests/custom_actions_spec.rb [hunk 1]
```diff
@@ -220,4 +220,471 @@ def expect_result_to_have_custom_actions_hash
...
+    it "accept custom actions" do
+      api_basic_authorize
+
+      post api_cluster_url(nil, @resource), :params => gen_request(:button1, "key1" => "value1")
+
+      expect_single_action_result(:success => true, :message => /.*/, :href => api_cluster_url(nil, @resource))
+    end
+  end
+
+  describe "ExtManagementSystem" do
```
Comment (abellotti) [inline spec/requests/custom_actions_spec.rb]: would prefer if these reflect the api collection, i.e. Providers,  also if you could sort them in this file, would be easier to look for the related section.  Thanks !!

File: spec/requests/custom_actions_spec.rb [hunk 2]
```diff
@@ -220,4 +220,471 @@ def expect_result_to_have_custom_actions_hash
...
+      get api_cloud_subnet_url(nil, @resource)
+
+      expect(response.parsed_body).to include(
+        "id"      => @resource.id.to_s,
+        "href"    => api_cloud_subnet_url(nil, @resource),
+        "actions" => a_collection_including(a_hash_including("name" => "button1"))
+      )
+    end
+
+    it "accept custom actions" do
```
Comment (jntullo) [inline spec/requests/custom_actions_spec.rb]: NIT, but I think all of these should say "accepts". I'll have to update my PRs as well.

File: spec/requests/custom_actions_spec.rb [hunk 3]
```diff
@@ -220,4 +220,471 @@ def expect_result_to_have_custom_actions_hash
...
+      api_basic_authorize
+
+      post api_user_url(nil, @resource), :params => gen_request(:button1, "key1" => "value1")
+
+      expect_single_action_result(:success => true, :message => /.*/, :href => api_user_url(nil, @resource))
+    end
+  end
+
+  describe "Vms" do
+    before(:each) do
```
Comment (jntullo) [inline spec/requests/custom_actions_spec.rb]: another NIT, probably just need `before`, not `before(:each)` (another thing I'll have to update in my own 😛 )

**Comment by jntullo** [comment]
Thanks for this @d-m-u !!! 🎉

File: spec/requests/custom_actions_spec.rb [hunk 4]
```diff
@@ -220,4 +220,471 @@ def expect_result_to_have_custom_actions_hash
...
+    it "accepts custom actions" do
+      api_basic_authorize
+
+      post api_load_balancer_url(nil, @resource), :params => gen_request(:button1, "key1" => "value1")
+
+      expect_single_action_result(:success => true, :message => /.*/, :href => api_load_balancer_url(nil, @resource))
+    end
+  end
+
+  describe "MiqGroup" do
```
Comment (abellotti) [inline spec/requests/custom_actions_spec.rb]: Group & sort.

File: spec/requests/custom_actions_spec.rb [hunk 5]
```diff
@@ -220,4 +220,471 @@ def expect_result_to_have_custom_actions_hash
...
+    it "accepts custom actions" do
+      api_basic_authorize
+
+      post api_group_url(nil, @resource), :params => gen_request(:button1, "key1" => "value1")
+
+      expect_single_action_result(:success => true, :message => /.*/, :href => api_group_url(nil, @resource))
+    end
+  end
+
+  describe "MiqTemplate" do
```
Comment (abellotti) [inline spec/requests/custom_actions_spec.rb]: Template & sort

**Comment by abellotti** [comment]
PR looks good @d-m-u, minor change needed.

@gmcculloug before we hit the merge button, being that it is marked g/yes, can anyone in the UI/SUI teams kick the tire ?

**Comment by gmcculloug** [comment]
@eclarizio Is probably best to test it out.

**Comment by abellotti** [comment]
🙇  @eclarizio when you get a chance can you kick the tires ?

**Comment by abellotti** [comment]
Just tried a query on container_nodes with 1000 items just expanding resources, the N+1 issue is there.  Before we merge these on master and g/yes, we have to fix the related n+1 issue with custom_actions generally(master and g/yes) otherwise we'll have a serious performance hit on 20+ collections.

**Comment by eclarizio** [comment]
Aside from @abellotti's performance concern, the custom actions being supported on all the new objects looks good to me.

**Comment by abellotti** [comment]
Thanks @eclarizio for testing them out 🙇

**Comment by abellotti** [comment]
@d-m-u I need to bother you to create a second PR 🙇 

As per https://github.com/ManageIQ/manageiq/pull/16467 not all of the collections targeted here are meant for the G-release, so the separate PR needs to be created for the collections meant to have custom_actions for the G-release.

these include:
- cloud_tenants
- clusters
- container_nodes
- data_stores
- hosts
- providers
- templates

Thank you.

/cc @gmcculloug

**Comment by d-m-u** [comment]
@jntullo or @abellotti can you give this and the counterpart https://github.com/ManageIQ/manageiq-api/pull/213 a 👀 just to make sure I didn't done goof on something silly once more please?

**Comment by abellotti** [comment]
Can't seem to find the custom_actions tests for these 8 collections.

**Comment by imtayadeway** [comment]
@abellotti before hitting the button on this (custom button?), have we considered the particularly thorny performance issues surrounding custom actions currently, i.e. https://github.com/ManageIQ/manageiq-api/issues/146 ?

**Comment by abellotti** [comment]
LGTM!! Thanks @d-m-u for updating this PR.  👍

**Pull request merged**